velox/common/memory/MmapAllocator.cpp

```diff
@@ -404,31 +447,111 @@ bool MmapAllocator::SizeClass::allocateLocked(
     }
     uint64_t bits = pageAllocated_[cursor];
     if (bits != kAllSet) {
-      if (considerMappedOnly > 0) {
-        uint64_t mapped = pageMapped_[cursor];
-        uint64_t mappedFree = ~bits & mapped;
-        if (mappedFree == 0) {
-          continue;
+      int previousToGo = numPagesToGo;
+      allocateAny(cursor, numPagesToGo, *numUnmapped, out);
+      numAllocatedUnmapped_ += previousToGo - numPagesToGo;
+      if (numPagesToGo == 0) {
+        clockHand_ = previousCursor;
+        return true;
+      }
+    }
+  }
+}
+
+namespace {
+bool isAllZero(xsimd::batch<uint64_t> bits) {
+  return simd::allSetBitMask<uint64_t>() ==
+      simd::toBitMask(bits == xsimd::broadcast<uint64_t>(0));
+}
+} // namespace
+
+int32_t MmapAllocator::SizeClass::findMappedFreeGroup() {
+  constexpr int32_t kWidth = 4;
+  int32_t index = lastLookupIndex_;
+  if (index == kNoLastLookup) {
+    index = 0;
+  }
+  auto lookupSize = mappedFreeLookup_.size() + kSimdTail;
+  for (auto counter = 0; counter <= lookupSize; ++counter) {
+    auto candidates = xsimd::load_unaligned(mappedFreeLookup_.data() + index);
+    auto bits = simd::allSetBitMask<int64_t>() ^
+        simd::toBitMask(candidates == xsimd::broadcast<uint64_t>(0LL));
+    if (!bits) {
+      index = index + kWidth <= mappedFreeLookup_.size() - kWidth
+          ? index + kWidth
+          : 0;
+      continue;
+    }
+    lastLookupIndex_ = index;
+    auto wordIndex = count_trailing_zeros(bits);
+    auto word = mappedFreeLookup_[index + wordIndex];
+    auto bit = count_trailing_zeros(word);
+    return (index + wordIndex) * 64 + bit;
+  }
+  ClassPageCount ignore = 0;
+  checkConsistency(ignore, ignore);
+  LOG(ERROR) << "MMAPL: Inconsistent mapped free lookup class " << unitSize_;
```

Comment:
To your comments somewhere else but related here:

> Generally inconsistencies here are a process killing situation because these signal some intractable corruption even though the gizmo can keep running for a long time.

So here the checkConsistency is called in runtime while doing allocation, instead of during testing. You also mentioned time of detection is not a factor. Wouldn't it be better to fail fast for easier debugging? This kind of issues cannot be detected by ASAN because we are using our own mapped memory. Would like to know your thoughts.